File: 0047 - Permutations II/solution.cpp
class Solution {
private:
    void solve(int idx, vector<int> A, vector<vector<int>>& res){
        if(idx == A.size()) {res.emplace_back(A); return;}
        for(int i = idx; i < A.size(); i++){
            if(i!=idx && A[i]==A[idx]) continue;
            swap(A[i], A[idx]);
            solve(idx+1, A, res);
        }
    }
public:
    vector<vector<int>> permuteUnique(vector<int>& A) {
        vector<vector<int>> res;
        sort(A.begin(), A.end());
        solve(0, A, res);
        return res;
    }
};

File: 0047 - Permutations II/solution.py
class Solution:
    def permuteUnique(self, A: List[int]) -> List[List[int]]:
        res, n = [], len(A)
        def solve(idx):
            if idx == n:
                res.append(A.copy()); return
            s = set()
            for i in range(idx, n):
                if A[i] not in s:
                    s.add(A[i])
                    A[idx], A[i] = A[i], A[idx]
                    solve(idx+1)
                    A[idx], A[i] = A[i], A[idx]
        solve(0)
        return res